### PR TITLE
Use ko to render full image specifications into the operator bundle image

### LINKED ISSUE #48: OperatorHub requires `containerImage` annotation
In the ClusterServiceVersion manifest, OperatorHub requires a `containerImage` annotation that points to the operator container image [1]. We should ensure that the manifest here contains this information and syncs it.

[1] https://operator-framework.github.io/community-operators/packaging-required-fields/

/kind feature

### PR DESCRIPTION
**Pull request by adambkaplan**
# Changes

- Move main.go to top-level, to conform with Kubebuilder v3 project
  format.
- Replace image references with ko replacements
- Remove kustomization that overrides image name and tag.
- Add containerImage to CSV base, using ko reference.
- Update Makefile to deploy operator directly with ko.
- Update Makefile to render the operator bundle CSV with ko.
- Update Makefile to have installation of ko idempotent.
- Update Makefile to deploy operator directly with ko.
- Update Makefile to have installation of ko idempotent.
- Update `make bundle-build` to build the OLM bundle in an output
  directory, using `ko resolve` to render the true CSV.
- Update `make bundle push` to also push the operator image.
- Update CI and release script to only run `make bundle-push`.
- Update development docs to drop references to `make ko-publish`.
- Print diff if check-git-status fails.

Fixes #48 

/kind feature

# Submitter Checklist

- [ ] Includes tests if functionality changed/was added
- [x] Includes docs if changes are user-facing
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [x] Release notes block has been filled in, or marked NONE

See [the contributor guide](https://github.com/shipwright-io/build/blob/master/CONTRIBUTING.md)
for details on coding conventions, github and prow interactions, and the code review process.

# Release Notes

```release-note
Use ko to render full image specifications into the operator bundle image. Add the `containerImage` annotation to the operator's CSV.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -41,6 +41,17 @@ IMAGE_TAG_BASE ?= $(IMAGE_REPO)/operator
...
 TAG ?= $(VERSION)
 
+# IMAGE_PUSH indicates if any of the images should be pushed. By default, images are not pushed
+# unless a command indicates that a push occurs.
+IMAGE_PUSH ?= false
+
+# Format for the SBOM produced by ko.
+# Defaults to "spdx", use "none" to disable SBOM generation
+SBOM ?= "spdx"
+
```
Comment (imjasonh) [inline Makefile:51]: ko may change its default ~soon, to either cyclonedx, or (more likely) both. Just something to keep in mind here.

File: Makefile [hunk 2]
```diff
@@ -217,22 +225,29 @@ rm -rf $$TMP_DIR ;\
...
 .PHONY: bundle-build
-bundle-build: bundle ## Build the bundle image.
-	$(CONTAINER_ENGINE) build -f bundle.Dockerfile -t $(BUNDLE_IMG) .
+bundle-build: bundle ko ## Build the bundle image. If IMAGE_PUSH is set to true, this will also build and push the operator image.
+	rm -rf _output/olm
+	mkdir -p _output/olm
+	cp -r bundle _output/olm/
+	cp bundle.Dockerfile _output/olm/
+	KO_DOCKER_REPO=${IMAGE_REPO} $(KO) resolve --push=${IMAGE_PUSH} ${KO_OPTS} -f ${OPERATOR_CSV} > _output/olm/${OPERATOR_CSV}
+	$(CONTAINER_ENGINE) build -f _output/olm/bundle.Dockerfile -t $(BUNDLE_IMG) _output/olm
```
Comment (imjasonh) [inline Makefile:246]: Sad that OLM undoes ko's hard work and requires a container runtime to make this bundle. Without knowing much about it I'd guess it's entirely unnecessary.

Comment (adambkaplan) [reply Makefile:246]: It doesn't undo ko's hard work, per se. The bundle is just YAML in an OCI artifact, which are updated with the `ko resolve`d image shas. A full runtime like docker/podman is indeed overkill, but in terms of what developers reasonably have available, the decision makes sense.

**Comment by adambkaplan** [comment]
One thing that this change can enable is nightly/latest releases for the operator! Since the bundle in git doesn't need to be changed for us to build a new bundle image, we can cut operator releases with a floating tag that respects semver.

**Comment by adambkaplan** [comment]
/approve

Self-approving

File: docs/development/olm-development.md [hunk 3]
```diff
@@ -9,28 +9,25 @@ Additional steps need to be taken to ensure the operator can be deployed with OL
...
-Run `make ko-publish IMAGE_REPO=<your-registry>`, pushing to a container registry that is accessible inside your Kubernetes cluster.
-Using `ko.local` or `kind.local` is not recommended, as this will not push the resulting image to a container registry.
+Run `make bundle-push IMAGE_REPO=<your-registry>`, pushing to a container registry that is
+accessible inside your Kubernetes cluster.
+This make command will push the operator and [operator bundle](https://olm.operatorframework.io/docs/tasks/creating-operator-bundle/)
+to the container registry.
+An operator bundle is an OCI artifact that tells OLM how to deploy your operator.
 
+Using `ko.local` or `kind.local` for `IMAGE_REPO` is not recommended, as this will not push the
+resulting images to an OCI-compliant container registry.
```
Comment (otaviof) [inline docs/development/olm-development.md:25]: That's an interesting behavior, @adambkaplan. I find the `ko` approach to `kind.local` registries very useful, but for OLM we really need OCI compliance.

**Review by otaviof — APPROVED** [review]
/lgtm

Thanks for the detailed documentation!

File: docs/development/olm-development.md (hunk 3 above)
Comment (adambkaplan) [reply docs/development/olm-development.md:25]: Yeah, found this the hard way when building the CI infra for the operator in the first place. Deploying an operator on OLM requires at minimum 3 container images:

1. The operator binary
2. The operator "bundle"
3. The catalog source

Item 3 needs to be pullable from a container registry. I _think_ item 2 must also be pullable from a container registry. In terms of dev/testing, pushing everything to a "real" registry just makes life easier.

**Pull request merged**